lib/SILOptimizer/SemanticARC/CoroutineLifetimeExtender.cpp

```diff
@@ -0,0 +1,295 @@
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+///
+/// This pass contains a simple transform that extends the lifetime of a
+/// coroutine with a trivial resume to transitive uses of copies of yielded
+/// values from the coroutine and then eliminates those copies.
+///
+//===----------------------------------------------------------------------===//
+
+#include "Context.h"
+#include "OwnershipLiveRange.h"
+#include "Transforms.h"
+
+#include "swift/SIL/ApplySite.h"
+#include "swift/SIL/BasicBlockUtils.h"
+#include "swift/SIL/OwnershipUtils.h"
+#include "swift/SIL/SILBuilder.h"
+#include "swift/SIL/SILInstruction.h"
+#include "swift/SILOptimizer/PassManager/Passes.h"
+#include "swift/SILOptimizer/PassManager/Transforms.h"
+#include "swift/SILOptimizer/Utils/ValueLifetime.h"
+
+using namespace swift;
+using namespace swift::semanticarc;
+
+//===----------------------------------------------------------------------===//
+//                        Process Candidate Functions
+//===----------------------------------------------------------------------===//
+
+static bool processCandidates(Context &ctx,
+                              ArrayRef<FullApplySite> applySiteList) {
+  assert(!applySiteList.empty() && "Must have at least one candidate");
+
+  SmallVector<SILInstruction *, 32> scratchBuffer;
+  SmallVector<SILInstruction *, 32> frontier;
+  SmallVector<SILInstruction *, 32> destroyInsts;
+  bool madeChange = false;
+
+  // Now optimize each apply site that we can (converting to guaranteed) and
+  // then after we are done, expand the lifetime of the coroutine over those
+  // points.
+  for (auto &fas : applySiteList) {
+    // Use RAII to ensure in the face of future code that we properly clear
+    // scratchBuffer, frontier on every iteration.
+    SWIFT_DEFER {
+      scratchBuffer.clear();
+      frontier.clear();
+      destroyInsts.clear();
+    };
+
+    // NOTE: This needs to be rewritten to use FullApplySite methods if we add
+    // additional coroutine call sites.
+    auto *bai = cast<BeginApplyInst>(fas);
+    bool didOptimizeCopy = true;
+
+    // Now we know that we have a simple enough coroutine that we can process,
+    // try to gather up copies that we /could/ eliminate if we lifetime extend.
+    // Return true if we found something and false otherwise.
+    for (SILValue yieldedValue : bai->getYieldedValues()) {
+      if (yieldedValue.getOwnershipKind() != ValueOwnershipKind::Guaranteed) {
+        continue;
+      }
+
+      bool result = visitGuaranteedValueUsesIgnoringForwarding(
+          yieldedValue, [&](Operand *use) {
+            auto *cvi = dyn_cast<CopyValueInst>(use->getUser());
+            if (!cvi) {
+              return true; // Continue processing.
+            }
```

Comment:
I think it's fine to add braces to single-line if statements if you want, but not if it also means adding two-lines of whitespace between the if statement and following statement. It's getting ridiculous if our standard `dyn_cast` idiom now takes 5 lines.